### DISCUSSION TITLE
"Wing Level Hold".

### DISCUSSION THREAD

**Discussion by laasmooi** [iNavFlight/inav#9269] [Q&A]
Hi all Inavers, 
I often use "Altitude Hold" but it should be nice to also have "Wing Level Hold". The idea is to climb quick and easy to a certain altitude while the wing stays level with the horizon but still is manually overridable.
Any ideas?

**Comment by Jetrell** [comment]
Do you mean a Constant Climb Angle function ?
So you can pull back on the elevator stick to set a certain climb angle, then release the stick. And the fixed wing will continue to climb at that angle. e.g. if you're flying up a long mountainous incline.

**Comment by MrD-RC** [comment]
Just use course hold mode, and add up elevators with a slider or pot. 

BTW, althold is not something that’s needed on fixed wing anymore. At this point in time, it should only be used with Angle mode.

**Reply by laasmooi** [reply]
Ah, that makes sense! Thanks.
It means when the plane/FC  is in course mode, I still have throttle and elevator control? And I can manually override the course?

**Reply by MrD-RC** [reply]
Yes. Course hold has manual throttle and you can change course.

**Comment by laasmooi** [comment]
That is all I need to know. Thanks!

**Comment by breadoven** [comment]
Well theoretically you could just use ACRO to do this. It should hold whatever attitude you've set the plane to for the climb, roll level, pitched to the desired climb angle. Unfortunately it tends to drift after a while so would need correcting every now and again to keep it on course.

**Reply by Jetrell** [reply]
>Well theoretically you could just use ACRO to do this

I've tried it with the integral on pitch and roll increased into the thirties.. As you said, it does eventually roll-off in one direction or the other.

This function could be useful, on long gradual climbs.
I was thinking of trying it in the programming logic... And making a Climb mode based on Cruise mode. Using Operation 45. With pitch angle constraints. 

![Capture](https://github.com/iNavFlight/inav/assets/47995726/5062ea8a-ea1d-480f-8d87-3fbc7a155ff5)

And using the elevator stick to set the required climb or dive angle. Then after 1s of releasing the stick to neutral, it would hold that attitude until the elevator stick was moved >1% either side of neutral, as a means to quickly disengage the climb, back to altitude holding.

**Reply by breadoven** [reply]
I've never checked how ACRO works exactly in detail but you'd think it saves the attitude state when the stick is released and then actively maintains the pitch and roll attitudes. But it drifts way off from the initial attitude and makes no attempt to correct so it must be doing something else.

**Reply by breadoven** [reply]
So after checking again ACRO just holds 0 rates when the stick is centered. I HITL tested a crude change for planes to make "ACRO" hold attitude instead, pretty much what `Flight_Axis_Angle_Override `does but the target angle is set to whatever the pitch/roll attitudes were when the roll/pitch stick was released back to centre (Is this still ACRO or some other mode in fact ?). Seems to work OK on the face of it but needs better logic controlling when it's active especially when you move outside the set pitch and roll bank angle limits. Also need to check the current ACRO again to see how it performs at lower airspeeds since that's when problems occur with drift, the plane simply doesn't have enough airspeed to maintain stability and drifts. With the change I made it tries to return to the last set attitudes when speed increases which is different to the current ACRO which just maintains the attitude it found itself in.

**Reply by MrD-RC** [reply]
To me, that’s more attitude hold. Acro is just a stabilised manual. Not to say that it couldn’t be introduced. But I’d still keep acro as it is now.

**Reply by Jetrell** [reply]
> Is this still ACRO or some other mode in fact ?

It depends if AIR mode is enabled.. It brings ACRO mode closer to the way an AVCS (Angular Velocity Control System) works..
But Acro/Air mode still lacks the software computations this type of controller uses to hold heading or attitude, based on the IMU alone.
The AVCS tail lock controller works very well for 3D Heli's, drift Cars and Planes.. Spektrum has had the AS3X system for years. And many of the Eagle Tree products did as well.
What you stated is true for fixed wing aircraft. An attitude lock controller is only as good as the thrust/airspeed being provided by the motor.
`fw_iterm_throw_limit = 165` was originally designed to prevent I-term windup before launch. Its also is a limiting factor to how well INAV will hold attitude on any axis. Especially when the control surface is command to provide more a more throw, to maintain the required attitude at low airspeeds or with strong wind gusts.... But INAV also lacks a means to provide adequate fixed wing I-term suppression when required.

I flew 3D planes for years.. An in maneuvers like a  Torque Roll. you pulse the throttle in the high alpha position, as each axis rotates skywards to provide enough control surface authority to hold the maneuver .. I've often thought if this same thing could be done by an FC auto throttle, which had an attitude lock controller.. Because without thrust, its a recipe for a crash, to those who don't understand what such a control mode requires to work correctly.

>To me, that’s more attitude hold. Acro is just a stabilised manual. Not to say that it couldn’t be introduced. But I’d still keep acro as it is now.

A separate attitude control mode to ACRO would be a better idea, if it ever happened.. So it wouldn't confuse pilots.

**Reply by MrD-RC** [reply]
> in maneuvers like a Torque Roll. you pulse the throttle in the high alpha position, as each axis rotates skywards to provide enough control surface authority to hold the maneuver .. I've often thought if this same thing could be done by an FC auto throttle, which had an attitude lock controller.. Because without thrust, its a recipe for a crash

isn’t that more because these 3D planes are tractors. So the pulsing throttle is to provide airflow over the control surfaces? I don’t know if that would do anything for a pusher configuration.

**Reply by Jetrell** [reply]
@MrD-RC True, a tractor prop is always going to provide better control surface reaction than a pusher.. But in saying that. Wing area and thrust to weight also plays a big part..
I'm not sure if you have ever flown indoor 3D models ?  Years back I did for a while. These planes had collective pitch, with elevators and ailerons that made up 2/3s of the wing and stabilizer area. And you could fly these planes backwards, and maintain good control in an inverted hover.

**Reply by breadoven** [reply]
This attitude hold definitely seems useful. If you set it up so it's only active within the set Roll and Pitch max bank angles, not inverted, it works very well. Handy for doing large radius turns which ACRO struggles with due to drift. Just need to decide how to implement it, new mode or activated some other way perhaps as an option for ACRO behaviour.

**Reply by Jetrell** [reply]
Maximum pitch bank angle can go up to 85 degrees... Which will require vigilance by the pilot to prevent a stall. 
That may get close to the point gimbal lock could be a problem.

**Reply by breadoven** [reply]
I'm thinking this is actually just ANGLE HOLD rather than ACRO and in fact it just uses the Angle code to work.

One way of using it without requiring a new mode would be to have it activate if Angle is selected when the Roll/Pitch stick is **not** centred. This would probably need a setting to control this behaviour. So if you want Angle to self level you select it with roll/pitch centred, otherwise if the setting is ON and Angle selected with the roll/pitch stick off centre, at the desired attitude in fact, then attitude hold becomes active. This should be OK for fixed wing which is the intended use right now. There is also the option to use Horizon as well of course anyway which would work separately. I personally never used Angle or Horizon on a fixed wing for self levelling because Altitude Hold does the same thing anyway (which uses auto enabled Angle).

As for speed issues I don't see this as being much different to ACRO. It's up to the pilot to check what's happening. Although you can always add a stick pusher function based on speed dropping too low which would return to level flight if need be.

**Reply by Jetrell** [reply]
Yes it would appear that `Flight_Axis_Angle_Override` is just ANGLE mode, in which the user can change what angle is LEVEL to the ground at center stick. Rather than it being an Attitude controller.

After reading your post, and what is actually required.. Having a selection switch etc, to engage and **escape** from the set attitude, with **safety**.. I think an extension of ANGLE mode itself, is likely to confuse users. Especially if they have to bail in a hurry to prevent a crash.
I really think it would be better to have it as LEVEL mode, or something like that, built upon a navigation controlled mode, like Cruise or Althold for fixedwings... So auto throttle could assist maintaining large radius turns or slow gradual climbs.

This way the user can hold the stick (roll or pitch) at the desired attitude for a second or so. Then it would hold that attitude.
And once the stick was released back to center. If it was moved a second time (outside a small deadband) it would disengage back to what the default of the mode it was built upon.. i.e holding altitude or level to the ground stabilization.
I don't think it would require more than 30degrees of pitch or roll. Considerably less than is required for good rate control over ANGLE mode.. Anymore would produce gradual side slip. Or push the wing past stall for a prolonged period of time in a climb, wasting energy.


> I personally never used Angle or Horizon on a fixed wing for self levelling because Altitude Hold does the same thing anyway (which uses auto enabled Angle).

Me neither.. Its either Acro or Cruise/ Course hold... But. there are a great many users that do use ANGLE mode.. Which is what I believe Darren was referring when he said it may be better to have it as a seperate mode.

**Reply by MrD-RC** [reply]
Oh, 100% I think it should be a new mode. Pilots are used to what is available already. This would just be like a more locked Acro, from a pilots perspective. From a code perspective, it may be basically angle with a variable level (the current level when the stick is centred). But from the pilot perspective, it’s closer to Acro, as there’s no self levelling and no rotational limits. 

So Attitude Hold as a mode would make most sense to me. I think that adding it to Angle with specific stick inputs would just be confusing. It could lead to panic crashes. I also don’t think the existing modes should be changed or renamed.

**Reply by breadoven** [reply]
You're right, a new mode is the best way. Just seems there are too many modes and not enough channels to control them although in this case I guess it would only be active when most other modes weren't, a bit like ACRO, so maybe it's not so difficult to configure things using an existing channel.

**Reply by Jetrell** [reply]
Maybe it could even be a flight mode modifier like Heading hold. That could be enabled with modes like Acro, Altitude hold, Course hold or Cruise, according to the users flight requirements ?

**Reply by MrD-RC** [reply]
Channels aren’t really a problem. All my 8 flight modes are on channel 6. It’s not a problem with OpenTX, EdgeTX, or ETHOS.

**Comment by laasmooi** [comment]
I'm just an ordinary user but I'm heavily impressed by your knowledge.
A good reason to study a bit more and hopefully understand this matter.
A big thanks anyway!

**Comment by sensei-hacker** [comment]
> Well theoretically you could just use ACRO to do this. ... Unfortunately it tends to drift 

Yep, this is just the definition of acro. Would just need get rid of the unintentional drift at center stick.

@MrD-RC 
>  This would just be like a more locked Acro, from a pilots perspective

Yeah just acro without the unwanted drift, right.

>  Oh, 100% I think it should be a new mode

Would you think a new mode should be used if we can instead just make acro behave as already defined - don't drift when at center stick?

I proposed a simple change to get rid of the drift, so center stick maintains attitude just as it's supposed to.


I think the answer is clear, but just to double check that you would agree. Should we have *these* two modes?:
Acro with uncommanded drift
Acro without unwanted drift

I would think that if we can get rid of the unwanted drift at center stick, there's no need for an extra mode. Would you agree?

@breadoven made a proposal which in turn gave me an idea on how to completely eliminate drift in rate mode, so it maintains attitude consistently when the stick is centered. In a simple way too.

**Reply by MrD-RC** [reply]
I don’t believe acro is supposed to be attitude lock. It’s just supposed to stabilise wind buffeting. Effectively manual with stability.

**Reply by sensei-hacker** [reply]
Maybe I'm misunderstanding. What is acro supposed the do when the sticks are centered?

**Reply by Jetrell** [reply]
ACRO applies a mild attitude stabilization, being based on a RATE controller, with the strenght of the i-term dictating how well it will attempt to hold attitude at center stick.
But in the case of Breadovens commit... Its different again... It uses the ANGLE rate controller, based on gravity vector level control offset. Which uses the AHRS via sensors to correct for attitude drift. To the best it can anyway, with some issues being noticed in my tests, beyond what the AHRS can compensate for at higher climb angles, when acceleration forces are close to 1G. (power train permitting)

**Comment by sensei-hacker** [comment]
> The idea is to climb quick and easy to a certain altitude while the wing stays level with the horizon but still is manually overridable.

That sounds like Angle mode or cruise, using your pitch trim if you want to be be completely hands off, be able to set your radio down while it climbs.

That's how we do long climb outs in full-sized aircraft.

**Reply by MrD-RC** [reply]
Totally agree. For a consistent climb. The best mode to use would be course hold. Then add up elevator with a pot. Acro with attitude hold isn’t what I would think of for this at all. That’s way too ArduPilot thinking. Acro is for having fun.
